src/x/resource/lifetime.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package resource
+
+import "sync"
+
+// CancellableLifetime describes a lifetime for a resource that
+// allows checking out the resource and returning it and once
+// cancelled will not allow any further checkouts.
+type CancellableLifetime struct {
+	mu        sync.RWMutex
+	cancelled bool
+}
+
+// NewCancellableLifetime returns a new cancellable resource lifetime.
+func NewCancellableLifetime() *CancellableLifetime {
+	return &CancellableLifetime{}
+}
+
+// TryCheckout will try to checkout the resource, if the lifetime
+// is already cancelled this will return false, otherwise it will return
+// true and guarantee the lifetime is not cancelled until the checkout
+// is returned.
+func (l *CancellableLifetime) TryCheckout() bool {
```

Comment:
Can you add to the comment: If this returns true you MUST call `ReleaseCheckout` later, but if it returns false calling `ReleaseCheckout` will panic (pretty sure unlocking an unlocked lock panics

Reply:
Yeah sure thing.

Reply:
Added comments.